### ISSUE TITLE
How are conf threshold and nms (iou) threshold working in segmentation task?

### ISSUE THREAD

**Issue by dimka11** [ultralytics/ultralytics#6967]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Are they only for detected boxes or for detected masks too?

**Comment by glenn-jocher** [comment]
@dimka11 the confidence (conf) threshold is a value that filters out detections based on the confidence score predicted by the model. For segmentation tasks, this thresholding generally applies to the candidate segments or masks, much like it does for bounding boxes in detection tasks.

The NMS (Non-Maximum Suppression) IoU (Intersection over Union) threshold primarily helps to resolve overlapping bounding boxes in object detection. In segmentation, if instance segmentation is in play, which produces separate masks for each instance of an object, IoU thresholds may be used to differentiate between overlapping instances.

In short, both thresholds work similarly by filtering and refining the results, though their application in segmentation may vary, particularly for the IoU threshold in the context of instance segmentation. For more task-specific insights, feel free to dive into our segmentation docs. 📚😊

**Comment by nitec427** [comment]
Hey @glenn-jocher I'm trying to use an instance segmentation model on my custom dataset, which then will be used as a CoreML model for further things. Just like in detection, I want to apply NMS to the model result to obtain wxh masks and visualize the results in Xcode Simulator. However, in export.py of yours NMS created some issues and I couldn't obtain a proper model. Idk, is there anyone come across the problem. It'd be good to have custom YOLO segmentations model that works in iOS systems. Here is the model properties that I desire (original DeepLab model presented in CoreML page) and the output of current model, respectively. Many thanks
<img width="839" alt="image" src="https://github.com/ultralytics/ultralytics/assets/23635935/7259c2ad-c451-4489-91f8-011e565cd658">
<img width="959" alt="image" src="https://github.com/ultralytics/ultralytics/assets/23635935/a2af451b-ee2b-4e7c-a7a1-f99b211d2a26">

**Comment by glenn-jocher** [comment]
@nitec427 hey there! For instance segmentation models, applying NMS is indeed a crucial step to refine the output masks. If you're encountering issues with NMS during the export process, it's possible that there might be a compatibility issue with the CoreML export pathway. We're always working on improving the export functionality, so your feedback is valuable.

For now, I recommend checking the export documentation and ensuring that all steps are followed correctly. If the problem persists, please open an issue with detailed error logs and the steps to reproduce it. This will help us pinpoint the issue and provide a fix or guidance. Thanks for bringing this to our attention, and we're keen on supporting the community in deploying YOLO segmentations on various platforms, including iOS. 🛠️😊

**Comment by nitec427** [comment]
Hi again @glenn-jocher. We have solved the problem, and seen that the problem is not related with YOLO. There were some issues during the conversion of our torchscript export file to CoreML. If anyone tries to do so, even this situation may arise from pytorch or CoreML not supporting even indexing like the following. [6:] We have to change this with torch.arange(6, len(<your_array>)) etc. Thanks you for always quickly answering our questions. I wish all the best.

**Comment by glenn-jocher** [comment]
@nitec427 Great to hear that you've resolved the issue! Your insight about the indexing incompatibility during the TorchScript to CoreML conversion is indeed valuable for others who might face similar challenges. Thank you for sharing your solution with the community, and don't hesitate to reach out if you have more questions in the future. Wishing you the best with your project! 😊🚀
